pennylane/transforms/gradients/qubit_parameter_shift.py

```diff
@@ -0,0 +1,114 @@
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains functions for computing the parameter-shift gradient
+of a qubit quantum tape.
+"""
+# pylint: disable=protected-access
+import numpy as np
+
+import pennylane as qml
+
+
+def expval_grad(tape, idx, gradient_recipe=None):
+    r"""Generate the parameter-shift tapes and postprocessing methods required
+    to compute the gradient of an expectation value with respect to an
+    expectation value.
+
+    Args:
+        tape (.QuantumTape): quantum tape to differentiate
+        idx (int): trainable parameter index to differentiate with respect to
+        gradient_recipe (tuple(Union(list[list[float]], None)) or None: Gradient recipe for the
+            parameter-shift method.
+
+            This is a tuple with one nested list per operation parameter. For
+            parameter :math:`\phi_k`, the nested list contains elements of the form
+            :math:`[c_i, a_i, s_i]` where :math:`i` is the index of the
+            term, resulting in a gradient recipe of
+
+            .. math:: \frac{\partial}{\partial\phi_k}f = \sum_{i} c_i f(a_i \phi_k + s_i).
+
+            If ``None``, the default gradient recipe containing the two terms
+            :math:`[c_0, a_0, s_0]=[1/2, 1, \pi/2]` and :math:`[c_1, a_1,
+            s_1]=[-1/2, 1, -\pi/2]` is assumed for every parameter.
+
+    Returns:
+        tuple[list[QuantumTape], function]: A tuple containing the list of generated tapes,
+        in addition to a post-processing function to be applied to the evaluated
+        tapes.
+
+    **Gradients of expectation values**
+
+    For a variational evolution :math:`U(\mathbf{p}) \vert 0\rangle` with
+    :math:`N` parameters :math:`\mathbf{p}`,
+
+    consider the expectation value of an observable :math:`O`:
+
+    .. math::
+
+        f(\mathbf{p})  = \langle \hat{O} \rangle(\mathbf{p}) = \langle 0 \vert
+        U(\mathbf{p})^\dagger \hat{O} U(\mathbf{p}) \vert 0\rangle.
+
+
+    The gradient of this expectation value can be calculated using :math:`2N` expectation
+    values using the parameter-shift rule:
+
+    .. math::
+
+        \frac{\partial f}{\partial \mathbf{p}} = \frac{1}{2\sin s} \left[ f(\mathbf{p} + s) -
+        f(\mathbf{p} -s) \right].
+
+    """
+    # check if the quantum tape contains any variance measurements
+    var_mask = [m.return_type is qml.operation.Variance for m in tape.measurements]
+
+    if any(var_mask):
+        raise ValueError("Does not support gradients of tapes with variance output.")
+
+    t_idx = list(tape.trainable_params)[idx]
+    op = tape._par_info[t_idx]["op"]
+    p_idx = tape._par_info[t_idx]["p_idx"]

```

Comment:
I guess this is copied and pasted, but that's quite some legacy code, no? Is this something we should flag for overhaul at some stage? Or does it still serve us well?

Reply:
I almost rewrote it, but I am often hesitant to refactor too much at once in case I introduce new bugs